llvm/lib/Transforms/Coroutines/SuspendCrossingInfo.h

```diff
@@ -0,0 +1,182 @@
+//===- SuspendCrossingInfo.cpp - Utility for suspend crossing values ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// The SuspendCrossingInfo maintains data that allows to answer a question
+// whether given two BasicBlocks A and B there is a path from A to B that
+// passes through a suspend point.
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIB_TRANSFORMS_COROUTINES_SUSPENDCROSSINGINFO_H
+#define LLVM_LIB_TRANSFORMS_COROUTINES_SUSPENDCROSSINGINFO_H
+
+#include "CoroInstr.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instruction.h"
+
+namespace llvm {
+
+// Provides two way mapping between the blocks and numbers.
+class BlockToIndexMapping {
+  SmallVector<BasicBlock *, 32> V;
+
+public:
+  size_t size() const { return V.size(); }
+
+  BlockToIndexMapping(Function &F) {
+    for (BasicBlock &BB : F)
+      V.push_back(&BB);
+    llvm::sort(V);
```

Comment:
Why is this sorting by pointer value?

Reply:
Then we can get the index of the BB by lower_bound. Otherwise we have to introduce a map here.

Reply:
But the block index is already available in BB.getNumber()?

Reply:
I didn't know this. Could we get the BB back from the number? If yes, this is an improvement.

Reply:
You can, but it requires chasing the block list. Is this used in a way that introduces runtime pointer value dependence?

Reply:
> You can, but it requires chasing the block list. 

So we need a linear search? Then it is not wanted.

>  Is this used in a way that introduces runtime pointer value dependence?

I am not sure I understand your points 100%. But this is only used to have a simple way to offer a two way map from BB and indexes. Then we can have a bitmap for the BB. And this is only used as an analysis in CoroSplit pass. So it at least won't introduce dependence on runtime pointer value across passes.

Reply:
I did a quick check for the use of this info in CoroSplit pass. And if I didn't miss a point, we won't create or remove BB during the use of the information. But I admit it is dangerous indeed.

Reply:
This is very interesting, I didn't notice this while reading the code earlier. The restriction that we cannot add/remove BB while using SuspendCrossingInfo is not clearly documented. I can add a comment for this.

I suppose if people don't mind using a map we can create one for BB.getNumber() -> BB*, but that is for another PR. Why is it a problem to use a map here?